drivers/network/dd/b57xx/ndis.c

```diff
@@ -0,0 +1,565 @@
+    
+    FreeSharedMemory(Adapter,
+                     sizeof(B57XX_SEND_BUFFER_DESCRIPTOR) * Adapter->SendProducer[0].Count,
+                     (PVOID)Adapter->SendProducer[0].pRing,
+                     &Adapter->SendProducer[0].RingControlBlock.HostRingAddress);
+    
+    if (Adapter->SendProducer[0].pPacketList != NULL)
+    {
+        NdisFreeMemory(Adapter->SendProducer[0].pPacketList,
+                       sizeof(PNDIS_PACKET) * Adapter->SendProducer[0].Count,
+                       0);
+    }
+    
+    if (Adapter->Status.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATUS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Status.pBlock,
+                              Adapter->Status.HostAddress);
+    }
+    
+    if (Adapter->Statistics.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATISTICS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Statistics.pBlock,
+                              Adapter->Statistics.HostAddress);
+    }
+    
+    /* Free MMIO */
+    if (Adapter->IoBase)
+    {
+        NdisMUnmapIoSpace(Adapter->MiniportAdapterHandle, Adapter->IoBase, Adapter->IoLength);
+    }
+
+    /* Destroy the adapter context */
+    NdisFreeMemory(Adapter, sizeof(B57XX_ADAPTER), 0);
+}
+
+NDIS_STATUS
+NTAPI
+MiniportSend(_In_ NDIS_HANDLE MiniportAdapterContext,
+             _In_ PNDIS_PACKET Packet,
+             _In_ UINT Flags)
+{
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    PSCATTER_GATHER_LIST SGList = NDIS_PER_PACKET_INFO_FROM_PACKET(Packet,
+                                                                   ScatterGatherListPacketInfo);
+    LONG BusyDescriptors = Adapter->SendProducer[0].ProducerIndex -
+                           Adapter->SendProducer[0].ConsumerIndex;
+    
+    NdisAcquireSpinLock(&Adapter->SendLock);
+    
+    ASSERT(SGList != NULL);
+    
+    if (BusyDescriptors < 0)
+    {
+        BusyDescriptors += Adapter->SendProducer[0].Count;
+    }
+    
+    if (Adapter->SendProducer[0].RingFull == TRUE ||
+        BusyDescriptors + SGList->NumberOfElements > Adapter->SendProducer[0].Count)
+    {
+        /* Force NDIS to resubmit the send */
+        return NDIS_STATUS_RESOURCES;
+    }
+    
+    for (UINT i = 0; i < SGList->NumberOfElements; i++)
+    {
+        ASSERT(SGList->Elements[i].Length <= Adapter->MaxFrameSize);
+        
+        Adapter->SendProducer[0].pPacketList[Adapter->SendProducer[0].ProducerIndex] = Packet;
+
+        NICTransmitPacket(Adapter,
+                          SGList->Elements[i].Address,
+                          SGList->Elements[i].Length,
+                          i == SGList->NumberOfElements - 1 ? B57XX_SBD_PACKET_END : 0);
```

Comment:
The number of elements is checked on every loop iteration. Why not just make the following:
```
  for (i = 0; i < SGList->NumberOfElements; i++)
  {
    Descriptor = Adapter->SendProducer[0].pRing[Index];
    Descriptor.Length = ...;
    Descriptor.Flags = ...;
    Descriptor.HostAddress = ...;

    Index = (Index + 1) % Adapter->SendProducer[0].Count;
  }
  Descriptor.Flags |= B57XX_SBD_PACKET_END;

  B57XXWriteMailbox(Adapter, B57XX_MBOX_TXP_HOST_RING_INDEX1, Index);
```

Reply:
As far as I know, other NDIS miniport drivers assert the number of elements to be 1. I added support for multiple elements to try and increase Windows Server 2003 compatibility. I don't believe it's necessary to restructure the send functions just to avoid a comparison. Though, I did move checking if the ring is full to `MiniportSend`

Reply:
Please restructure the send functions.
See nvnet code for example:
https://github.com/reactos/reactos/blob/master/drivers/network/dd/nvnet/send.c#L104